### PR TITLE
Add `legacyBundle` config option to pages & exported-components in BM Flow

### LINKED ISSUE #2527: Multi-GAables in flow-bm
# Summary 

Since many of our users are already / want to split their applications into smaller projects, we want to support this use-case in the BM flow.

There are 2 general ways we've seen users splitting their BM applications into separately-GAable artifacts:

1. ### Single module bundle

<details>
<summary><code>module_MODULE_ID.json</code>:</summary>

```json
{
	"...": "...",
	"bundles": [
		{
			"file": {
        		"artifactId": "com.wixpress.some-root-bm-module",
		        "path": "module.bundle.min.js"
			},
			"debugFile": { "...": "..." }
    	}
	]
}
```
</details>

Where `some-root-bm-module`'s `module.bundle.js` bundle registers everything BM-related:

<details>
<summary><code>module.ts</code>@<code>some-root-bm-module</code>:</summary>

```typescript
class MyModule extends BusinessManagerModule {
	constructor(moduleId) {
		super(moduleId);
		this.registerPageComponent(
			'some-page-a', 
			SomePageALazyWrapper // imported from npm from what I've seen
		);
		this.registerPageComponent(
			'some-page-b', 
			SomePageBLazyWrapper // same
		);
		// ...
	}
}
```
</details>

2. ### Multiple module bundles

<details>
<summary><code>module_MODULE_ID.json</code>:</summary>

```json
{
	"...": "...",
	"bundles": [
		{
			"file": {
        		"artifactId": "com.wixpress.bm-module-a",
		        "path": "module.bundle.min.js"
			},
			"debugFile": { "...": "..." }
    	},
		{
			"file": {
        		"artifactId": "com.wixpress.bm-submodule-b",
		        "path": "module.bundle.min.js"
			},
			"debugFile": { "...": "..." }
    	},
		{ "...": "..." }
	]
}
```
</details>

Where every artifact's `module.bundle.js` bundle registers everything BM-related.

<details>
<summary><code>module.ts</code>@<code>bm-module-a</code>:</summary>

```typescript
class ModuleA extends BusinessManagerModule {
	constructor(moduleId) {
		super(moduleId);
		this.registerPageComponent(
			'some-page-a', 
			SomePageLazyWrapper // imported locally
		);
	}
}
```
</details>

<details>
<summary><code>module.ts</code>@<code>bm-submodule-b</code>:</summary>

```typescript
class SubModuleB extends BusinessManagerModule {
	constructor(moduleId) {
		super(moduleId);
		this.setModuleConfigurationId('bm-module-a');
		this.registerPageComponent(
			'some-page-b', 
			SomePageLazyWrapper // imported locally
		);
	}
}
```
</details>

## Context
Since Business-Manager themselves recommend the 2nd approach, and it allows for changes in submodules which don't require a GA on the "main" module - we chose this approach for BM flow.

## The Issue
How do we enable users who chose approach 1 (2 didn't exist before), to gradually migrate their GAables to BM flow without first shifting their approach from no. 1 to no. 2?

## Proposed Solution
Add a config option to pages & exported-components `legacyBundleName` which when supplied, will (also) build a dedicated bundle for the page/exported-component, named after the provided config option.
This page can then be referred to the same way it was before.
This (unless I missed something) allows for a seamless transition between the 2 modes.

### PR DESCRIPTION
**Pull request by amitdahan**
### 🔦 Summary
Closes #2527.

This PR adds a `legacyBundle` config option to pages & exported-components, for example -

- `src/pages/index.json`
```json
{
	"legacyBundle": {
		"bundleName": "bundle-name",
		"lazyComponentId": "component-id"
	}
}
```

Will create an additional bundle for the `src/pages/index.tsx` page component, which will register the component to the `ModuleRegistry` using the supplied `legacyBundle.lazyComponentId`.

Using this config option can help with transitioning into the new way of encapsulating separate GAables into a dedicated BM module bundle (which will be added to the app's module json file).

### CODE REVIEW

File: packages/yoshi-flow-bm/src/renderLegacyWrappers.ts [hunk 1]
```diff
@@ -0,0 +1,69 @@
+import fs from 'fs-extra';
```
Comment (amitdahan) [inline packages/yoshi-flow-bm/src/renderLegacyWrappers.ts:1]: Dead code please remove

File: packages/yoshi-flow-bm/src/config/validConfig.ts [hunk 2]
```diff
@@ -25,10 +25,18 @@ export const validModuleConfig: InitialModuleConfig = {
...
+    lazyComponentId: 'some-legacy-page-component-id',
+  },
 };
 
 export const validExportedComponentConfig: InitialExportedComponentConfig = {
   componentId: 'component-id',
+  legacyBundle: {
+    bundleName: 'some-legacy-bundle-name',
+    lazyComponentId: 'some-legacy-component-id',
+  },
```
Comment (yanivefraim) [inline packages/yoshi-flow-bm/src/config/validConfig.ts:39]: super small - we can probably reuse this code

File: packages/yoshi-flow-bm/src/renderExportedComponent.ts [hunk 3]
```diff
@@ -34,24 +34,36 @@ export default wrapComponent(Component, [
...
+  }
+  ${
+    addFedops
+      ? `createFedopsProvider(${JSON.stringify(component.componentId)}),`
+      : ''
+  }
+]);`;
 };
 
+export const EXPORTED_COMPONENT_ENTRY = ({
```
Comment (yanivefraim) [inline packages/yoshi-flow-bm/src/renderExportedComponent.ts:50]: why is it in capital letters? 😄

Comment (ranyitz) [inline packages/yoshi-flow-bm/src/renderExportedComponent.ts:50]: This shouldn't be uppercase

File: packages/yoshi-flow-bm/src/renderLegacyExportedComponent.ts [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+import WrappedComponent from '${EXPORTED_COMPONENT_ENTRY(exportedComponent)}';
+import { ModuleRegistry } from 'react-module-container';
+
+ModuleRegistry.registerComponent('${
+    exportedComponent.config.legacyBundle!.lazyComponentId
+  }', () => WrappedComponent);
+`;
+};
+
+const EXPORTED_COMPONENT_LEGACY_ENTRY = ({
```
Comment (ranyitz) [inline packages/yoshi-flow-bm/src/renderLegacyExportedComponent.ts:21]: Shouldn't be UPPER_CASED

File: packages/yoshi-flow-bm/src/renderLegacyExportedComponent.ts [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+    )
+    .reduce<Entry>((entries, exportedComponent) => {
+      entries[
+        exportedComponent.config.legacyBundle?.bundleName!
+      ] = EXPORTED_COMPONENT_LEGACY_ENTRY(exportedComponent);
+      return entries;
+    }, {});
+};
+
+export default renderLegacyExportedComponent;
```
Comment (ranyitz) [inline packages/yoshi-flow-bm/src/renderLegacyExportedComponent.ts:54]: Maybe better to change the file name to `legacyExportedComponent` and have this as a named export

File: packages/yoshi-flow-bm/src/scripts/start.ts [hunk 6]
```diff
@@ -96,7 +98,11 @@ const start: CliCommand = async function (argv, yoshiConfig) {
     isDev: true,
     isHot: yoshiConfig.hmr as boolean,
   });
-  clientConfig.entry = getModuleEntry(model);
+  clientConfig.entry = {
```
Comment (ranyitz) [inline packages/yoshi-flow-bm/src/scripts/start.ts:101]: Getting the entry can be extracted into another function (so it won't be duplicated between start and build)

File: packages/yoshi-flow-bm/src/renderLegacyPage.ts [hunk 7]
```diff
@@ -0,0 +1,36 @@
+import path from 'path';
+import fs from 'fs-extra';
+import { Entry } from 'webpack';
+import { GENERATED_LEGACY_DIR, PAGES_DIR } from './constants';
+import { FlowBMModel, PageModel } from './model';
+import { PAGE_ENTRY } from './renderPage';
+
+const generatePageLegacyCode = (page: PageModel) => {
```
Comment (yavorsky) [inline packages/yoshi-flow-bm/src/renderLegacyPage.ts:8]: Maybe we can consider sharing `t` util between editor flow and bm flow? It allows to have typed templates and call it as a functions.

Comment (yavorsky) [reply packages/yoshi-flow-bm/src/renderLegacyPage.ts:8]: For example: https://github.com/wix/yoshi/blob/master/packages/yoshi-flow-editor/src/wrappers/templates/CommonEditorScriptEntry.ts#L19

Comment (amitdahan) [reply packages/yoshi-flow-bm/src/renderLegacyPage.ts]: Sounds great, long overdue.
Will push a separate PR for this 💪

**Pull request merged**